src/arr/compiler/resolve-scope.arr

```diff
@@ -118,23 +119,147 @@ fun desugar-toplevel-types(stmts :: List<A.Expr>) -> List<A.Expr> block:
   end
 end
 
+is-s-contract = A.is-s-contract
+type Contract = A.Expr%(is-s-contract)
+
+# Note: binds are maintained in reversed order, for efficiency in adding new items to them
+# They get reversed back to correct order in bind-wrap.
 data BindingGroup:
-  | let-binds(binds :: List<A.LetBind>)
-  | letrec-binds(binds :: List<A.LetrecBind>)
+  | let-binds(contracts :: List<Contract>, binds :: List<A.LetBind>)
+  | letrec-binds(contracts :: List<Contract>, binds :: List<A.LetrecBind>)
   | type-let-binds(binds :: List<A.TypeLetBind>)
 end
 
+var errors = empty # THE MUTABLE LIST OF ERRORS
+
+fun weave-contracts(contracts, rev-binds) block:
+  # When weaving contracts, ensure that the contract location is before the definition
+  # or else give a well-formedness error.
+  # NOTE(Ben): This code is polymorphic over LetBind and LetrecBind
+  contracts-sd = [SD.mutable-string-dict: ]
+  for each(c from contracts):
+    name = c.name.toname()
+    cases(Option) contracts-sd.get-now(name):
+      | none =>
+        contracts-sd.set-now(name, c)
+      | some(c2) =>
+        errors := link(C.contract-redefined(c.l, name, c2.l), errors)
+    end
+  end
+  fun rebuild-bind(bind, new-b, new-v):
+    if      A.is-s-let-bind(bind): A.s-let-bind(bind.l, new-b, new-v)
+    else if A.is-s-var-bind(bind): A.s-var-bind(bind.l, new-b, new-v)
+    else if A.is-s-letrec-bind(bind): A.s-letrec-bind(bind.l, new-b, new-v)
+    end
+  end
+  fun names-match(funargs :: List<A.Bind>, annargs :: List<A.AField>):
+    if is-empty(funargs) and is-empty(annargs): true
+    else if is-empty(funargs) or is-empty(annargs): false
+    else:
+      (funargs.first.id.toname() == annargs.first.name) and names-match(funargs.rest, annargs.rest)
```

Comment:
Underscores don't matter here: we're weaving the annotations onto the argument names before we desugar away underscores.